core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java

```diff
@@ -518,7 +573,7 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
             if (shard.unassigned()) {
                 continue;
             }
-            if (shard.primary() == false) {
+            if (shard.primary() == false || shard.relocating()) { // for relocating primaries, we replica from target to source if source is marked as relocated
                 nodesSentTo.remove(shard.currentNodeId());
             }
             if (shard.relocating()) {
```

Comment:
can we add an assertion that we actually removed something? I think this will also mean that we need a tighter control over what we remove. For example, if the primary is relocating, but we did not move to "relocated" mode, the currentNodeId should not be in the list (and shouldn't be removed).